src/Components/Components/src/Rendering/Renderer.cs

```diff
@@ -398,13 +398,22 @@ private ComponentState GetOptionalComponentState(int componentId)
                 ? componentState
                 : null;
 
-        private void ProcessRenderQueue()
+        /// <summary>
+        /// Processses pending renders requests from components.
+        /// </summary>
+        protected virtual void ProcessRenderQueue()
```

Comment:
Rather than making this part of the supported API surface, could we keep `ProcessRenderQueue` as `private` here, and instead add something like:

```cs
protected virtual bool AllowStartingNewRenderBatch { get => true; }
```

... so that subclasses can optionally put in their own logic for that flag? Then we would leave `ProcessRenderQueue` alone, and instead change the last bit of `AddToRenderQueue` to look like:

```cs
            if (!_isBatchInProgress && AllowStartingNewRenderBatch)
            {
                ProcessRenderQueue();
            }
```

This way we know that we're *only* changing the behavior with respect to when things ask to be rendered and would start a new batch. It means we're not changing the behavior about how batches are built once they already start being processed, which potentially avoids a lot of complexity (we rely on the fact that once a batch gets started, any time a parent component tries to render a child, we *must* do so synchronously because the array for the parameters would be returned to the pool synchronously anyway).

Reply:
We need to be able to trigger renders explicitly from subclasses, so ProcessRenderQueue is a cleaner way to achieve it. For the record what you are suggesting was my initial approach.

I do get the argument about the public API surface, but at the same time no one other than us is going to subclass the renderer. (In blazor server-side its internal and you don't get to replace it, same as on web assembly)

If we are concerned about this, we should simply move the entire renderer class to the .Internal namespace

Reply:
> at the same time no one other than us is going to subclass the renderer

That's definitely not the case. There are multiple other teams actively doing this today (well, I know of two). The `Renderer` base does need to be public, and its API surface needs to be robust and supported.

> We need to be able to trigger renders explicitly from subclasses

OK, right - understood. But I'm still keen to ensure that this can't be misused through confusion, and that this new feature only affects *components asking for themselves to be rendered*, not the way that batches-already-in-progress are processed.

So how about:

```cs
private bool _allowNewRenderBatches = true;

protected void PauseRenderBatches()
{
    _allowNewRenderBatches = false;
}

protected void ResumeRenderBatches()
{
    _allowNewRenderBatches = true;
    if (!_isBatchInProgress && _batchBuilder.ComponentRenderQueue.Count > 0)
    {
        ProcessRenderQueue();
    }
}

protected internal virtual void AddToRenderQueue(int componentId, RenderFragment renderFragment)
{
        // ... same as before, except changing the ending to:

        if (!_isBatchInProgress && _allowNewRenderBatches)
        {
            ProcessRenderQueue();
        }
}
```

I'd find that a lot easier to reason about, because now we know this feature only changes the timing of when we issue a new batch, and doesn't mean that batches can get stopped halfway through.

Reply:
Alternatively, if you prefer:

```
protected virtual bool DeferNewRenderRequests { get => false; }

protected void ProcessRenderQueue() // protected but *not* virtual
{
    if (_isBatchInProgress)
    {
        throw new InvalidOperationException("A batch is already in progress. ProcessRenderQueue should only be called from a quiescent state.");
    }

    if (_batchBuilder.ComponentRenderQueue.Count == 0)
    {
        return;
    }

    // ... all the existing logic here ...
}

protected internal virtual void AddToRenderQueue(int componentId, RenderFragment renderFragment)
{
        // ... same as before, except changing the ending to:

        if (!(_isBatchInProgress || DeferNewRenderRequests))
        {
            ProcessRenderQueue();
        }
}
```

This does make `ProcessRenderQueue` callable by subclasses, but only when it's safe to do so, and avoids interfering with how we process batches already in flight.

Reply:
> OK, right - understood. But I'm still keen to ensure that this can't be misused through confusion, and that this new feature only affects _components asking for themselves to be rendered_, not the way that batches-already-in-progress are processed.

Could we simply make ProcessRenderQueue run inside the sync context (that way there's no chance calling it multiple times can affect existing render batches) or have a check inside it that says:
```
if(_isBatchInProgress) 
{
    return;
}
```

Reply:
I called EnsureSynchronizationContext as part of ProcessRenderQueue and returned directly if there were no components to render. I think that's a better choice for the following reasons:
* Makes it impossible to misuse ProcessRenderQueue.
  * It forces it to execute inside the sync context explicitly (instead of implicitly as it was before through DispatchEvent or AddToRenderQueue (both of which checked this).
  * It effectively serializes the batch processing (we didn't change anything here, as we expect them to be serialized).
* Adds a single public API method instead of a combination of methods/flags.
* You can simply call ProcessRenderQueue when you need it (invoking it inside the dispatcher) and it'll.
  * Render a batch from the components that requested a render.
  * Noop.